README.md
Oauth2 Yii2 integration
=======================

This extension allow the developper to use [Oauth2](https://bshaffer.github.io/oauth2-server-php-docs/) server.

[![Latest Stable Version](https://poser.pugx.org/sweelix/yii2-oauth2-server/v/stable)](https://packagist.org/packages/sweelix/yii2-oauth2-server)
[![Build Status](https://api.travis-ci.org/pgaultier/yii2-oauth2.svg?branch=master)](https://travis-ci.org/pgaultier/yii2-oauth2)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/pgaultier/yii2-oauth2/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/pgaultier/yii2-oauth2/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/pgaultier/yii2-oauth2/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/pgaultier/yii2-oauth2/?branch=master)
[![License](https://poser.pugx.org/sweelix/yii2-oauth2-server/license)](https://packagist.org/packages/sweelix/yii2-oauth2-server)

[![Latest Development Version](https://img.shields.io/badge/unstable-devel-yellowgreen.svg)](https://packagist.org/packages/sweelix/yii2-oauth2-server)
[![Build Status](https://travis-ci.org/pgaultier/yii2-oauth2.svg?branch=devel)](https://travis-ci.org/pgaultier/yii2-oauth2)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/pgaultier/yii2-oauth2/badges/quality-score.png?b=devel)](https://scrutinizer-ci.com/g/pgaultier/yii2-oauth2/?branch=devel)
[![Code Coverage](https://scrutinizer-ci.com/g/pgaultier/yii2-oauth2/badges/coverage.png?b=devel)](https://scrutinizer-ci.com/g/pgaultier/yii2-oauth2/?branch=devel)
[![composer.lock](https://poser.pugx.org/sweelix/yii2-oauth2-server/composerlock)](https://packagist.org/packages/sweelix/yii2-oauth2-server)

Installation
------------

If you use Packagist for installing packages, then you can update your composer.json like this :

``` json
{
    "require": {
        "sweelix/yii2-oauth2-server": "~1.2.0"
    }
}
```

Howto use it
------------

Add extension to your configuration

``` php
return [
    //....
    'bootstrap' => [
        //....
        'oauth2',
        //....
    ],
    'modules' => [
        //....
        'oauth2' => [
            'class' => 'sweelix\oauth2\server\Module',
            'backend' => 'redis',
            'db' => 'redis',
            'identityClass' => 'app\models\User', // only if you don't want to use the user identityClass
            //
            // Parameters
            //
        ],
        //....
    ],
    //....
];
```

Configure Module
----------------

### Basic module parameters

 * `backend` : can only be **redis** for the moment
 * `db` : id of the redis component or connection or connection configuration
 * `identityClass` : user class used to link oauth2 authorization system default to user component `identityClass`
 * `webUserParamId` : allow separation between main app user (session) and module app user, (default to **__oauth2**)
 * `identityCookieName` : allow separation between main app user (cookie) and module app user, (default to **oauth2**)
 * `webUser` : allow full management of module web user, (default to **[]**)
 * `baseEndPoint` : base path for token and authorize endpoints default to `''`
    * Token endpoint https://host.xxx/token
    * Authorize endpoint https://host.xxx/authorize
 * `overrideLayout` : override module layout to use another one (ex: @app/views/layouts/oauth2)
 * `overrideViewPath` : override view path to use specific one (ex: @app/views/oauth2)   

### Grants management 
 
 * `allowImplicit` : allow implicit grant (default to **false**)
 * `allowAuthorizationCode` : allow authorization code grant (default to **true**)
 * `allowClientCredentials` : allow client credentials grant (default to **true**)
 * `allowPassword` : allow user credentials / password grant (default to **true**)
 * `allowCredentialsInRequestBody` : allow credentials in request body (default to **true**)
 * `allowPublicClients` : allow public clients (default to **true**)
 * `alwaysIssueNewRefreshToken` : always issue refresh token (default to **true**)
 * `unsetRefreshTokenAfterUse` : unset refresh token after use (default to **true**) 

### JWT parameters

 * `allowJwtAccessToken` : enable JWT (default : **false**)
 * `allowAlgorithm` : available algorithm for JWT (default : **['RS256', 'RS384', 'RS512']**)
 * `jwtAudience` : default to token endpoint
 * `storeEncryptedTokenString` : store encrypted token (default : **true**)

### Time To Live

 * `idTTL` : TTL of ID Token (default to **3600**)
 * `accessTokenTTL` : TTL of access token (default to **3600**)
 * `refreshTokenTTL` : TTL of refresh token (default to **14 * 24 * 3600**)

### Basic Oauth names

 * `realm` : Realm value (default to **Service**)
 * `tokenQueryName` : name of the access token parameter (default to **access_token**)
 * `tokenBearerName` : name of authorization header (default to **Bearer**)

### Enforce parameters
 
 * `enforceState` : enforce state parameter (default to **true**)
 * `allowOnlyRedirectUri` : need exact redirect URI (default to **true**)

### OpenID

 * `allowOpenIdConnect` : enable openId connect (default : **false**) // not implemented yet

### Authorization Code parameters

 * `enforceRedirect` : enforce redirect parameter (default to **false**)
 * `authorizationCodeTTL` : TTL of authorization code (default to **30**)

### CORS

 * `cors` : enable `CORS` on the token endpoint (default : **false**) the CORS part can be defined using an array as described [in Yii documentation](http://www.yiiframework.com/doc-2.0/yii-filters-cors.html)
 
``` php
 return [
     //....
     'bootstrap' => [
         //....
         'oauth2',
         //....
     ],
     'modules' => [
         //....
         'oauth2' => [
             'class' => 'sweelix\oauth2\server\Module',
             'backend' => 'redis',
             'db' => 'redis',
             'identityClass' => 'app\models\User', // only if you don't want to use the user identityClass
             //
             // Cors parameters example :
             //
             'cors' => [
                'Origin' => ['https://www.myowndomain.com'],
             ]
         ],
         //....
     ],
     //....
 ];

```
 
User identity and Web user
--------------------------

Configure the user component to link oauth2 system and user / identity management

``` php
return [
    //....
    'components' => [
        //....
        'user' => [
            'class' => 'sweelix\oauth2\server\web\User',
            'identityClass' => 'app\models\User', // Identity class must implement UserModelInterface
            //
            // Parameters
            //
        ],
        //....
    ],
    //....
];
```

`IdentityClass` must implements `sweelix\oauth2\server\interfaces\UserModelInterface`. You can use the trait
`sweelix\oauth2\server\traits\IdentityTrait` to automagically implement 

 * `public function getRestrictedScopes()`
 * `public function setRestrictedScopes($scopes)`
 * `public static function findIdentityByAccessToken($token, $type = null)`

you will have to implement the remaining methods : 

 * `public static function findByUsernameAndPassword($username, $password)`
 * `public static function findByUsername($username)`

Creating specific view for OAuth2
---------------------------------

In order to use your own views (instead of the builtin ones), you can override 
 * `layout` : module parameter `overrideLayout`
 * `viewPath` : module parameter `overrideViewPath`
 
### Overriding layout
 
You should create a classic layout like :
 
```php
<?php
/**
 * @app/views/layouts/newLayout.php
 * @var string $content
 */
use yii\helpers\Html;

$this->beginPage(); ?>
    <!DOCTYPE html>
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title><?php echo Html::encode($this->title); ?></title>

        <meta name="viewport" content="width=device-width, initial-scale=1">

        <?php $this->head(); ?>
    </head>
    <body>
        <?php $this->beginBody(); ?>
            <?php echo $content;?>
        <?php $this->endBody(); ?>
    </body>

</html>
<?php $this->endPage();

``` 

and link it to the module

```php
return [
    //....
    'modules' => [
        //....
        'oauth2' => [
            'class' => 'sweelix\oauth2\server\Module',
            'overrideLayout' => '@app/views/layouts/newLayout',
            //
            // Additional Parameters
            //
        ],
        //....
    ],
    //....
];
```

### Overriding views
 
You should create 3 views to allow oauth2 module to work as expected and link them to the module

```php
return [
    //....
    'modules' => [
        //....
        'oauth2' => [
            'class' => 'sweelix\oauth2\server\Module',
            // use views in folder oauth2
            'overrideViewPath' => '@app/views/oauth2',
            //
            // Additional Parameters
            //
        ],
        //....
    ],
    //....
];
```
 
#### Error view
 
This view is used to display a page when an error occurs
 
```php
<?php
/**
 * error.php
 *
 * @var string $type error type
 * @var string $description error description
 */
use yii\helpers\Html;
?>

    <h1 class="alert-heading"><?php echo ($type ? : 'Unkown error'); ?></h1>
    <p><?php echo ($description ? : 'Please check your request'); ?></p>

``` 


#### Login view
 
This view is used to display a login page when needed
 
```php
<?php
/**
 * login.php
 *
 * @var \sweelix\oauth2\server\forms\User $user
 *
 */
use yii\helpers\Html;
?>
    <?php echo Html::beginForm('', 'post', ['novalidate' => 'novalidate']); ?>
        <label>Username</label>
        <?php echo Html::activeTextInput($user, 'username', [
            'required' => 'required',
        ]); ?>
        <br/>
    
        <label>Password</label>
        <?php echo Html::activePasswordInput($user, 'password', [
            'required' => 'required',
        ]); ?>
        <br/>
        <button type="submit">LOGIN</button>
    <?php echo Html::endForm(); ?>

``` 

#### Authorize view
 
This view is used to display an authorization page when needed
 
```php
<?php
/**
 * authorize.php
 *
 * @var \sweelix\oauth2\server\interfaces\ScopeModelInterface[] $requestedScopes
 * @var \sweelix\oauth2\server\interfaces\ClientModelInterface $client
 *
 */
use yii\helpers\Html;
?>
    <h1><?php echo $client->name ?> <span>requests access</span></h1>
    
    <?php echo Html::beginForm(); ?>
        <?php if(empty($requestedScopes) === false) : ?>
        <ul>
            <?php foreach($requestedScopes as $scope): ?>
            <li>
                <h4><?php echo $scope->id; ?></h4>
                <p>
                    <?php echo $scope->definition; ?>
                </p>
            </li>
            <?php endforeach; ?>
        </ul>
        <?php endif; ?>
            <!-- name of decline button **must be** decline -->
            <button type="submit" name="decline">DECLINE</button>
            <!-- name of accept button **must be** accept -->
            <button type="submit" name="accept">AUTHORIZE</button>
    <?php echo Html::endForm(); ?>

``` 

Exposed Models overview
-----------------------

The Oauth2 Yii2 extension expose severall models which can be used in your application.
All models can be overloaded using Yii2 DI.

For example, if you want to overload the `Client` model, you have to inject your own model in the DI using:

```php

Yii::$container->set('sweelix\oauth2\server\interfaces\ClientModelInterface', [
    'class' => YourClientModel::className(),
]);
```

### Client / ClientModelInterface

 * `Client::findOne($id)` - Find client by ID
 * `Client::findAllByUserId($id)` - Find all clients accepted by user (userId)
 * `$client->save()` - Save client
 * `$client->delete()` - Delete client
 * `$client->hasUser($userId)` - Check if user (userId) has accepted the client
 * `$client->addUser($userId)` - Attach the user (userId) to the client
 * `$client->removeUser($userId)` - Dettach the user (userId) from the client
 
### AccessToken / AccessTokenModelInterface

 * `AccessToken::findOne($id)` - Find accessToken by ID
 * `AccessToken::findAllByUserId($id)` - Find all accessTokens for user (userId)
 * `AccessToken::findAllByClientId($id)` - Find all accessTokens for client (clientId)
 * `AccessToken::deleteAllByUserId($id)` - Delete all accessTokens for user (userId)
 * `AccessToken::deleteAllByClientId($id)` - Delete all accessTokens for client (clientId)
 * `$accessToken->save()` - Save accessToken
 * `$accessToken->delete()` - Delete accessToken

### RefreshToken / RefreshTokenModelInterface

 * `RefreshToken::findOne($id)` - Find accessToken by ID
 * `RefreshToken::findAllByUserId($id)` - Find all refreshTokens for user (userId)
 * `RefreshToken::findAllByClientId($id)` - Find all refreshTokens for client (clientId)
 * `RefreshToken::deleteAllByUserId($id)` - Delete all refreshTokens for user (userId)
 * `RefreshToken::deleteAllByClientId($id)` - Delete all refreshTokens for client (clientId)
 * `$refreshToken->save()` - Save refreshToken
 * `$refreshToken->delete()` - Delete refreshToken

### AuthCode / AuthCodeModelInterface

 * `AuthCode::findOne($id)` - Find authCode by ID
 * `$authCode->save()` - Save authCode
 * `$authCode->delete()` - Delete authCode

### Scope / ScopeModelInterface

 * `Scope::findOne($id)` - Find scope by ID
 * `Scope::findAvailableScopeIds()` - Find all scopes IDs
 * `Scope::findDefaultScopeIds()` - Find default scopes IDs
 * `$scope->save()` - Save scope
 * `$scope->delete()` - Delete scope

### CypherKey / CypherKeyModelInterface

 * `CypherKey::findOne($id)` - Find cypherKey by ID
 * `$cypherKey->save()` - Save cypherKey
 * `$cypherKey->delete()` - Delete cypherKey
 * `$cypherKey->generateKeys()` - Generate random keys for current cypherKey

Linking RBAC and Scope systems
------------------------------

Using `sweelix\oauth2\server\web\User` class will automagically link `rbac` system and `oauth2` system.

Permission system will be slightly modified to allow fine grained checks :

 * `Yii::$app->user->can('read')` will check
    1. if scope `read` is allowed for current client
    2. if rbac permission `read` is allowed for current user 
 
 * `Yii::$app->user->can('rbac:read')` will check **only** if rbac permission `read` is allowed for current user 

 * `Yii::$app->user->can('oauth2:read')` will check **only** if scope `read` is allowed for current client

Running the tests
-----------------

Before running the tests, you should edit the file tests/config/redis.php and change the config to match your environment.

CLI System
----------

Several commands are available to manage oauth2 system

 * `php protected/yii.php oauth2:client/create`
 * `php protected/yii.php oauth2:client/update`
 * `php protected/yii.php oauth2:key/create`
 * `php protected/yii.php oauth2:scope/create`


CHANGELOG.md
Oauth2 Yii2 Change Log
======================

1.3.0 May 11, 2018
------------------

 * Enh: update minimal Yii requirement to version 2.0.13
 * Enh: update console script deprecated error codes
 * Enh: update deprecated method from `Yii::trace()` to `Yii::debug()`
 * Enh: update deprecated method from `className()` to `::class`
 * Fix: make module PHP 7.2 compliant
 
1.2.1 July 11, 2017
-------------------

 * Chg: remove testing support for `HHVM` and php `5.6`
 * Chg: update dependencies
 * Enh: adding `.gitattributes`
 
1.2.0 April 10, 2017
--------------------

 * Chg: Update `fxp-asset`
 * Chg: update dependencies
 * Enh: Adding new methods `findAllByUserId()` and `findAllByClientId()` for models `AccessToken` and `RefreshToken`
 * Enh: Adding new methods `hasUser()`, `addUser()`, `removeUser()` and `findAllByUserId()` from model `Client`
 * Enh: Allow oauth2 `user` to be fully decoupled from app `user` 
 * Fix: Fix Jwt token
 * Enh: Enable redis automatic expire for `AccessToken`, `RefreshToken` and `AuthorizationCode`

1.1.0 January 02, 2017
----------------------

 * Chg: update dependencies
 * Enh: Adding `CORS` support for `token` endpoint
 * Enh: Allow `HttpBearerAuth` and `QueryParamAuth` for method `findIdentityByAccessToken` 

1.0.3 December 16, 2016
-----------------------

 * Enh: Allow multiple redirectUri creation in `oauth2:client/create` use `,` as separator
 * Fix: Correct Array conversion in `oauth2:client/create`
 * Fix: Remove userId type check
 * Fix: fix `allowJwtAccessToken` typo (was `allowJwtAccesToken`)
 * Enh: update composer.json to check openssl ext
 * Fix: Fix badges in readme
 * Enh: adding multiple redirectUri support.

1.0.2 December 6, 2016
----------------------

 * Chg: update dependencies
 * Enh: adding gitlab ci

1.0.1 November 7, 2016
----------------------

 * Fix: add missing view / layout override in module conf

1.0.0 November 7, 2016
----------------------

 * Fix: use yii2-redis 2.0.4 because 2.0.5 cannot reopen connection after closing it
 * Chg: Initial public release


src/Module.php
<?php
/**
 * Module.php
 *
 * PHP version 5.6+
 *
 * @author Philippe Gaultier <pgaultier@sweelix.net>
 * @copyright 2010-2017 Philippe Gaultier
 * @license http://www.sweelix.net/license license
 * @version 1.2.0
 * @link http://www.sweelix.net
 * @package sweelix\oauth2\server
 */
namespace sweelix\oauth2\server;

use sweelix\oauth2\server\services\Oauth;
use sweelix\oauth2\server\services\Redis;
use yii\base\BootstrapInterface;
use yii\base\Module as BaseModule;
use yii\console\Application as ConsoleApplication;
use Yii;
use yii\di\Instance;
use yii\helpers\ArrayHelper;

/**
 * Oauth2 server Module definition
 *
 * @author Philippe Gaultier <pgaultier@sweelix.net>
 * @copyright 2010-2017 Philippe Gaultier
 * @license http://www.sweelix.net/license license
 * @version 1.2.0
 * @link http://www.sweelix.net
 * @package sweelix\oauth2\server
 * @since 1.0.0
 */
class Module extends BaseModule implements BootstrapInterface
{
    /**
     * @var string backend to use, available backends are 'redis'
     */
    public $backend;

    /**
     * @var Connection|array|string the DB connection object or the application component ID of the DB connection.
     */
    public $db;

    /**
     * @var string override layout. For example @app/views/layouts/oauth2 to use <app>/views/layouts/oauth2.php layout
     */
    public $overrideLayout;

    /**
     * @var string override view path. For example @app/views/oauth2 to use <app>/views/oauth2/(authorize|login|error) views
     */
    public $overrideViewPath;

    /**
     * This user class will be used to link oauth2 authorization system with the application.
     * The class must implement \sweelix\oauth2\server\interfaces\UserInterface
     * If not defined, the Yii::$app->user->identityClass value will be used
     * @var string|array user class definition.
     */
    public $identityClass;

    /**
     * @var string used to separate user session between this module and current application
     */
    public $webUserParamId = '__oauth2';

    /**
     * @var string used to separate identity cookies between this module and current application
     */
    public $identityCookieName = 'oauth2';

    /**
     * @var array webUser configuration specific to this module
     */
    public $webUser = [];

    /**
     * @var string change base end point
     */
    public $baseEndPoint = '';

    /**
     * @var bool configure oauth server (use_jwt_access_tokens)
     */
    public $allowJwtAccessToken = false;

    /**
     * @var array configure oauth server (allowed_algorithms)
     */
    public $allowAlgorithm = ['RS256', 'RS384', 'RS512'];

    /**
     * @var string|array jwt audience. Default to token endpoint
     */
    public $jwtAudience = ['token/index'];

    /**
     * @var bool configure oauth server (store_encrypted_token_string)
     */
    public $storeEncryptedTokenString = true;

    /**
     * @var bool configure oauth server (use_openid_connect)
     */
    public $allowOpenIdConnect = false;

    /**
     * @var int configure oauth server (id_lifetime)
     */
    public $idTTL = 3600;

    /**
     * @var int configure oauth server (access_lifetime)
     */
    public $accessTokenTTL = 3600;

    /**
     * @var int configure oauth server (refresh_token_lifetime)
     */
    public $refreshTokenTTL = 1209600;

    /**
     * @var string configure oauth server (www_realm)
     */
    public $realm = 'Service';

    /**
     * @var string configure oauth server (token_param_name)
     */
    public $tokenQueryName = 'access_token';

    /**
     * @var string configure oauth server (token_bearer_header_name)
     */
    public $tokenBearerName = 'Bearer';

    /**
     * @var bool configure oauth server (enforce_state)
     */
    public $enforceState = true;

    /**
     * @var bool configure oauth server (require_exact_redirect_uri)
     */
    public $allowOnlyRedirectUri = true;

    /**
     * @var bool configure oauth server (allow_implicit)
     */
    public $allowImplicit = false;

    /**
     * @var bool allow authorization code grant
     */
    public $allowAuthorizationCode = true;

    /**
     * @var bool allow client credentials grant
     */
    public $allowClientCredentials = true;

    /**
     * @var bool allow password grant
     */
    public $allowPassword = true;

    /**
     * @var bool configure oauth server (allow_credentials_in_request_body)
     */
    public $allowCredentialsInRequestBody = true;

    /**
     * @var bool configure oauth server (allow_public_clients)
     */
    public $allowPublicClients = true;

    /**
     * @var bool configure oauth server (always_issue_new_refresh_token)
     */
    public $alwaysIssueNewRefreshToken = true;

    /**
     * @var bool configure oauth server (unset_refresh_token_after_use)
     */
    public $unsetRefreshTokenAfterUse = false;

    /**
     * @var int duration of login time for multiple authorize calls
     */
    public $loginDuration = 60 * 60 * 24 * 30;

    /**
     * @var bool configure authorization code (enforce_redirect)
     */
    public $enforceRedirect = false;

    /**
     * @var int configure authorization code (auth_code_lifetime)
     */
    public $authorizationCodeTTL = 30;

    /**
     * @var false|array Cors configuration if allowed @see http://www.yiiframework.com/doc-2.0/yii-filters-cors.html
     */
    public $cors = false;

    /**
     * @inheritdoc
     */
    public function init()
    {
        parent::init();
    }

    /**
     * Load dataservices in container
     * @param \yii\base\Application $app
     * @since 1.0.0
     */
    protected function setUpDi($app)
    {
        if (Yii::$container->has('scope') === false) {
            Yii::$container->set('scope', 'sweelix\oauth2\server\validators\ScopeValidator');
        }
        if (Yii::$container->has('sweelix\oauth2\server\interfaces\AccessTokenModelInterface') === false) {
            Yii::$container->set('sweelix\oauth2\server\interfaces\AccessTokenModelInterface', 'sweelix\oauth2\server\models\AccessToken');
        }
        if (Yii::$container->has('sweelix\oauth2\server\interfaces\AuthCodeModelInterface') === false) {
            Yii::$container->set('sweelix\oauth2\server\interfaces\AuthCodeModelInterface', 'sweelix\oauth2\server\models\AuthCode');
        }
        if (Yii::$container->has('sweelix\oauth2\server\interfaces\ClientModelInterface') === false) {
            Yii::$container->set('sweelix\oauth2\server\interfaces\ClientModelInterface', 'sweelix\oauth2\server\models\Client');
        }
        if (Yii::$container->has('sweelix\oauth2\server\interfaces\CypherKeyModelInterface') === false) {
            Yii::$container->set('sweelix\oauth2\server\interfaces\CypherKeyModelInterface', 'sweelix\oauth2\server\models\CypherKey');
        }
        if (Yii::$container->has('sweelix\oauth2\server\interfaces\JtiModelInterface') === false) {
            Yii::$container->set('sweelix\oauth2\server\interfaces\JtiModelInterface', 'sweelix\oauth2\server\models\Jti');
        }
        if (Yii::$container->has('sweelix\oauth2\server\interfaces\JwtModelInterface') === false) {
            Yii::$container->set('sweelix\oauth2\server\interfaces\JwtModelInterface', 'sweelix\oauth2\server\models\Jwt');
        }
        if (Yii::$container->has('sweelix\oauth2\server\interfaces\RefreshTokenModelInterface') === false) {
            Yii::$container->set('sweelix\oauth2\server\interfaces\RefreshTokenModelInterface', 'sweelix\oauth2\server\models\RefreshToken');
        }
        if (Yii::$container->has('sweelix\oauth2\server\interfaces\ScopeModelInterface') === false) {
            Yii::$container->set('sweelix\oauth2\server\interfaces\ScopeModelInterface', 'sweelix\oauth2\server\models\Scope');
        }
        if ((Yii::$container->has('sweelix\oauth2\server\interfaces\UserModelInterface') === false) && ($this->identityClass !== null)) {
            Yii::$container->set('sweelix\oauth2\server\interfaces\UserModelInterface', $this->identityClass);
        }
        if ($this->backend === 'redis') {
            Redis::register($app);
        }
        Oauth::register($app);

    }

    /**
     * @inheritdoc
     */
    public function bootstrap($app)
    {
        // use the registered identity class if not overloaded
        if (($this->identityClass === null) && (isset($app->user) === true)) {
            $this->identityClass = $app->user->identityClass;
        }
        $this->setUpDi($app);
        if (empty($this->baseEndPoint) === false) {
            $this->baseEndPoint = trim($this->baseEndPoint, '/').'/';
        }

        if ($app instanceof ConsoleApplication) {
            $this->mapConsoleControllers($app);
        } else {
            $app->getUrlManager()->addRules([
                ['verb' => 'POST', 'pattern' => $this->baseEndPoint.'token', 'route' => $this->id.'/token/index'],
                ['verb' => 'OPTIONS', 'pattern' => $this->baseEndPoint.'token', 'route' => $this->id.'/token/options'],
                ['verb' => 'GET', 'pattern' => $this->baseEndPoint.'authorize', 'route' => $this->id.'/authorize/index'],
                ['pattern' => $this->baseEndPoint.'authorize-login', 'route' => $this->id.'/authorize/login'],
                ['pattern' => $this->baseEndPoint.'authorize-application', 'route' => $this->id.'/authorize/authorize'],
                ['pattern' => $this->baseEndPoint.'authorize-error', 'route' => $this->id.'/authorize/error'],
            ]);
        }
    }

    /**
     * @inheritdoc
     */
    public function beforeAction($action)
    {
        $status = parent::beforeAction($action);
        // override web user to avoid conflicts only when routing into this module
        if ($status === true) {
            $userConfig = [
                'class' => 'yii\web\User',
                'identityClass' => $this->identityClass,
                'enableAutoLogin' => true,
                'enableSession' => true,
                'identityCookie' => ['name' => $this->identityCookieName, 'httpOnly' => true],
                'idParam' => $this->webUserParamId,
            ];
            $userConfig = ArrayHelper::merge($userConfig, $this->webUser);

            Yii::$app->set('user', $userConfig);
        }
        return $status;
    }

    /**
     * Update controllers map to add console commands
     * @param ConsoleApplication $app
     * @since 1.0.0
     */
    protected function mapConsoleControllers(ConsoleApplication $app)
    {
        $app->controllerMap['oauth2:client'] = [
            'class' => 'sweelix\oauth2\server\commands\ClientController',
        ];
        $app->controllerMap['oauth2:scope'] = [
            'class' => 'sweelix\oauth2\server\commands\ScopeController',
        ];
        $app->controllerMap['oauth2:key'] = [
            'class' => 'sweelix\oauth2\server\commands\KeyController',
        ];

    }
}
